File: android/app/src/main/java/com/safedrop/android/net/Protocol.kt
package com.safedrop.android.net

import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import org.json.JSONObject

private const val MAX_FRAME = 64 * 1024 * 1024  // 64 MB cap, matches Python.

object Protocol {
    fun sendFrame(out: OutputStream, payload: ByteArray) {
        require(payload.size <= MAX_FRAME) { "frame too large (${payload.size} bytes)" }
        val header = ByteArray(4)
        val len = payload.size
        header[0] = (len ushr 24).toByte()
        header[1] = (len ushr 16).toByte()
        header[2] = (len ushr 8).toByte()
        header[3] = (len and 0xFF).toByte()
        out.write(header)
        out.write(payload)
        out.flush()
    }

    fun recvFrame(input: InputStream): ByteArray {
        val header = ByteArray(4)
        readExact(input, header)
        val len =
            ((header[0].toInt() and 0xFF) shl 24) or
            ((header[1].toInt() and 0xFF) shl 16) or
            ((header[2].toInt() and 0xFF) shl 8) or
             (header[3].toInt() and 0xFF)
        require(len in 0..MAX_FRAME) { "frame too large ($len bytes)" }
        val payload = ByteArray(len)
        readExact(input, payload)
        return payload
    }

    fun sendJson(out: OutputStream, msg: JSONObject, encrypt: ((ByteArray) -> ByteArray)? = null) {
        val raw = msg.toString().toByteArray(Charsets.UTF_8)
        val payload = encrypt?.invoke(raw) ?: raw
        sendFrame(out, payload)
    }

    fun recvJson(input: InputStream, decrypt: ((ByteArray) -> ByteArray)? = null): JSONObject {
        val raw = recvFrame(input)
        val decoded = decrypt?.invoke(raw) ?: raw
        return JSONObject(decoded.toString(Charsets.UTF_8))
    }

    private fun readExact(input: InputStream, into: ByteArray) {
        var read = 0
        while (read < into.size) {
            val n = input.read(into, read, into.size - read)
            if (n < 0) throw EOFException("peer closed connection mid-frame")
            read += n
        }
    }
}

File: safedrop/protocol.py
"""JSON message framing for SafeDrop TCP control channel.

Wire format: [4-byte big-endian length N][N bytes payload].
Payload is JSON bytes — plaintext during the handshake, Fernet-encrypted
ciphertext afterwards. The framing layer itself is agnostic to that.
"""

from __future__ import annotations

import json
import socket
import struct
from typing import Any, Callable

from .config import LENGTH_PREFIX

MAX_FRAME = 64 * 1024 * 1024  # 64 MB — sanity cap to refuse absurd frames.


class ProtocolError(Exception):
    """Raised on malformed or oversized frames."""


def _recv_exact(sock: socket.socket, n: int) -> bytes:
    chunks: list[bytes] = []
    remaining = n
    while remaining:
        chunk = sock.recv(remaining)
        if not chunk:
            raise ConnectionError("peer closed connection mid-frame")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def send_frame(sock: socket.socket, payload: bytes) -> None:
    if len(payload) > MAX_FRAME:
        raise ProtocolError(f"frame too large ({len(payload)} bytes)")
    sock.sendall(struct.pack(">I", len(payload)) + payload)


def recv_frame(sock: socket.socket) -> bytes:
    header = _recv_exact(sock, LENGTH_PREFIX)
    (length,) = struct.unpack(">I", header)
    if length > MAX_FRAME:
        raise ProtocolError(f"frame too large ({length} bytes)")
    return _recv_exact(sock, length)


def send_json(sock: socket.socket, msg: dict[str, Any], encrypt: Callable[[bytes], bytes] | None = None) -> None:
    raw = json.dumps(msg, ensure_ascii=False).encode("utf-8")
    if encrypt is not None:
        raw = encrypt(raw)
    send_frame(sock, raw)


def recv_json(sock: socket.socket, decrypt: Callable[[bytes], bytes] | None = None) -> dict[str, Any]:
    raw = recv_frame(sock)
    if decrypt is not None:
        raw = decrypt(raw)
    try:
        msg = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProtocolError(f"invalid JSON frame: {exc}") from exc
    if not isinstance(msg, dict) or "type" not in msg:
        raise ProtocolError("JSON frame missing 'type' field")
    return msg
